Some typedefs are global and some aren't in --checkJs

**TypeScript Version:**  nightly (2.4.0-dev.20170516)

**Code**

These are excerpts that illustrate why it's weird:

dev-tools/globals.js

```js
/** @typedef {{[k: string]: any}} AnyObject */
```

sim/dex-data.js

```js
/** @typedef {{[k: string]: any}} AnyEffect */
```

sim/dex.js

```js
/** @type {AnyObject} */
let a; // this isn't an error
/** @type {AnyEffect} */
let b; // this IS an error
```

tsconfig.json

```js
{
    "include": [
        "./dev-tools/globals.ts",
        "./sim/dex-data.js",
        "./sim/dex.js"
    ]
}
```

These are excerpts that explain why this behavior is so weird, but the real reason why it works this way is probably in the rest of the code, which you can see for yourself in: https://github.com/Zarel/Pokemon-Showdown/tree/149ca3759c141085d2f473fc9fdc5da5371ef32c

**Expected behavior:**
No error, or perhaps two errors.

**Actual behavior:**
> sim/dex.js(3,12): error TS2304: Cannot find name 'AnyEffect'.

It's actually massively confusing to me how to access types defined in other files (I've googled and can't find any documentation on it). The general impression I get is that in modern TypeScript versions, in `.ts` files, you just export them and import them and they'll work normally, but still nothing on how to do it in `.js` files.

But either way, this seems like a bug.

(I tried moving the `AnyEffect` typedef from `sim/dex-data.js` to `dev-tools/globals.js` and I still get an error for it but not `AnyObject`, so there is almost certainly something weird going on here.)

--- a/src/compiler/parser.ts
+++ b/src/compiler/parser.ts
@@ -685,7 +685,7 @@ namespace ts {
 
             sourceFile.statements = parseList(ParsingContext.SourceElements, parseStatement);
             Debug.assert(token() === SyntaxKind.EndOfFileToken);
-            sourceFile.endOfFileToken = <EndOfFileToken>parseTokenNode();
+            sourceFile.endOfFileToken = addJSDocComment(parseTokenNode() as EndOfFileToken);
 
             setExternalModuleIndicator(sourceFile);
 

--- a/tests/baselines/reference/jsdocTypeDefAtStartOfFile.symbols
+++ b/tests/baselines/reference/jsdocTypeDefAtStartOfFile.symbols
@@ -1,0 +1,12 @@
+=== tests/cases/conformance/jsdoc/dtsEquivalent.js ===
+/** @typedef {{[k: string]: any}} AnyEffect */
+No type information for this code./** @typedef {number} Third */
+No type information for this code.=== tests/cases/conformance/jsdoc/index.js ===
+/** @type {AnyEffect} */
+let b;
+>b : Symbol(b, Decl(index.js, 1, 3))
+
+/** @type {Third} */
+let c;
+>c : Symbol(c, Decl(index.js, 3, 3))
+

--- a/tests/baselines/reference/jsdocTypeDefAtStartOfFile.types
+++ b/tests/baselines/reference/jsdocTypeDefAtStartOfFile.types
@@ -1,0 +1,12 @@
+=== tests/cases/conformance/jsdoc/dtsEquivalent.js ===
+/** @typedef {{[k: string]: any}} AnyEffect */
+No type information for this code./** @typedef {number} Third */
+No type information for this code.=== tests/cases/conformance/jsdoc/index.js ===
+/** @type {AnyEffect} */
+let b;
+>b : { [k: string]: any; }
+
+/** @type {Third} */
+let c;
+>c : number
+

--- a/tests/cases/conformance/jsdoc/jsdocTypeDefAtStartOfFile.ts
+++ b/tests/cases/conformance/jsdoc/jsdocTypeDefAtStartOfFile.ts
@@ -1,0 +1,11 @@
+// @allowJs: true
+// @checkJs: true
+// @noEmit: true
+// @Filename: dtsEquivalent.js
+/** @typedef {{[k: string]: any}} AnyEffect */
+/** @typedef {number} Third */
+// @Filename: index.js
+/** @type {AnyEffect} */
+let b;
+/** @type {Third} */
+let c;

